runtime_lib/aircpu/channel.cpp

```diff
@@ -0,0 +1,260 @@
+// Copyright (C) 2023, Advanced Micro Devices, Inc. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "air_tensor.h"
+
+#include <cstdint>
+#include <cstdio>
+
+#define VERBOSE 0
+
+/// Utility function to check if a memory address is valid
+bool isValidMemoryAddress(uint64_t address)
+{
+  try
+  {
+    volatile uint8_t *ptr = reinterpret_cast<volatile uint8_t *>(address);
+    volatile uint8_t value = *ptr; // Perform a memory access
+    return true;                   // If no exception occurs, the address is valid
+  }
+  catch (...)
+  {
+    return false; // Exception occurred, the address is invalid
+  }
+}
+
+template <typename T, int R>
+static void _air_channel_put(tensor_t<uint64_t, 0> *channel, tensor_t<T, R> *src,
+                             size_t *_offset, size_t *_size, size_t *_stride)
+{
+  size_t offset[4] = {0, 0, 0, 0};
+  size_t size[4] = {1, 1, 1, 1};
+  size_t stride[4] = {1, 1, 1, 1};
+  for (int i = 0; i < R; i++)
+  {
+    offset[i] = _offset[i];
+    size[i] = _size[i];
+    stride[i] = _stride[i];
+  }
+  // allocate a buffer in heap memory
+  T *buffer = (T *)malloc(sizeof(T) * size[3] * size[2] * size[0] * size[1]);
+  if (VERBOSE)
+    printf("dst offset %lu, %lu, size %lu, %lu, stride %lu, %lu\n", offset[1],
+           offset[0], size[1], size[0], stride[1], stride[0]);
```

Comment:
You love C. :-)
At least send debug/error message on `stderr` or use `std::cerr`?

Reply:
sorry, we'll spell it `std::printf` next time.